### ISSUE TITLE
Add support for ISOBMFF Files (AVIF, HEIF, CR3)

### ISSUE THREAD

**Issue by johnny-bit** [Exiv2/exiv2#1229]
**Is your feature request related to a problem? Please describe.**
Exiv2 does not support ISOBMFF format, therefore doesn't support AVIF, HEIF, CR3 and possibly other filetypes using this format

**Describe the solution you'd like**
Ideally - ISOBMFF format support, at least reading data from it :)

**Describe alternatives you've considered**
ExifTool and couple other open source projects already support ISOBMFF based files, however exiv2 still doesn't.

**Additional context**
Iam fully aware of #1066 - I asked my legal counsel and gethered much needed advice. According to all info I managed to gather, there's no legal issue in creating and using a tool capable of reading data and metadata stored in ISOBMFF containers. Worth noting is the fact that JPEG2000 IS an ISOBMFF based and support for it is present in the library without any issue.

Aditionally, already existing libraries like libavif, libheif, libraw and others already work with ISOBMFF files without legal problems.

The ONLY problematic issue would be not reading but decrypting encrypted (not encoded) data or using proprietary code, which I am sure is not the issue here.

**Comment by Filou83** [comment]
Please add support for CR3 and HEIF/HEIC image formats, at least to support reading the image/EXIF informations located in the container.

I've read #1066 , #978 and the comment here https://github.com/darktable-org/rawspeed/issues/121#issuecomment-636817674. I tried to understand the concerns about legal issues when reading data from patented ISOBMFF, but I could not see any evidence in this concerns. As far as I've read, there already have been contributions submitted to support these file fomats. Please add them to your library.

Thank you in advance!

**Comment by clanmills** [comment]
This would now be in Exiv2 v0.27.3 RC2 without this show stopper from @D4N https://github.com/Exiv2/exiv2/issues/1066#issuecomment-605429885

It's too late to develop the code for v0.27.3 which is now at RC2.  https://github.com/Exiv2/exiv2/issues/1018#issuecomment-604539346

I will support and mentor/assist anybody willing to work on this issue.

**Comment by Filou83** [comment]
> 
> 
> This would now be in Exiv2 v0.27.3 without this show stopper from @D4N [#1066 (comment)](https://github.com/Exiv2/exiv2/issues/1066#issuecomment-605429885)
> 
> It's too late to develop the code required to implement this feature for v0.27.3 which is now at RC2. [#1018 (comment)](https://github.com/Exiv2/exiv2/issues/1018#issuecomment-604539346)
> 
> I will support and mentor/assist anybody willing to work on this issue.

Thank you for your answer. But I still don't understand exactly what the goal is. There have been 2 contributions done to bring HEIF support to exiv2, maybe these could be reopened?
#978 and #1044 
Then #1066 has been opened. So there are 3 proposals to get ISOBMFF support. Maybe one of them could be pushed again into master branch? Maybe it could be your last huge step forward for the project before retireing from exiv2? ;) (with a winking eye and full of hope)
Or do you believe, the only possible option is to write independend code in exiv2 without using the libheif library?

**Comment by clanmills** [comment]
I'm working on a book "Image Metadata and Exiv2 Architecture".  Here's today's version:  https://clanmills.com/exiv2/book/  (and/or https://clanmills.com/exiv2/book/IMaEA.pdf).

The book is coming along quite steadily.  The purpose behind the book is to help Exiv2 survive and be developed in future.

It was not only Dan's comment, there was another very negative comment:  https://github.com/Exiv2/exiv2/issues/1066#issuecomment-609879359  I didn't understand the comment because it refers to a case about which I know nothing.

I don't think we need to use another library to do this.  We already have a CR2 and JP2000 parsers.  We can probably modify that and we're done.  And there's another brainless approach.  Exif metadata is written as an embedded TIFF which has the signature `II*\0` or `MM\0*`, XMP `<x:xmpmeta x="adobe:ns:meta/"`  We could read the file byte-by-byte and find the metadata.  What's the difference between reading a file byte-by-byte and using the command `cp`?  None that I can think of.  Does `cp` infringe image patents?

However, as Dan said "Patents don't work that way!", and I took that to mean "no matter what solution you propose, you may be in legal trouble!".  A couple of years ago, we needed legal advice and I got free help from an expert in intellectual property licensing.  I offered to consult with this person before publish any code for ISOBMFF.  There was no response to my offer.

Exiv2 v0.27.3 will ship as scheduled on 2020-06-30.  I'm going to finish the book and have offered to do "one final release" which will be 0.27.4 or 0.28.  Again, no feedback from the team.

We could have a team meeting (on Zoom or something) to discuss this.  I can't be bothered.

**Comment by pr0m1th3as** [comment]
@clanmills Thank you for all your work with exiv2 and please do everyone a big favor and go ahead with the initial plan of adding ISOBMFF support to your last "one final release". 

A few words about software patents and multimedia formats

Software patents are put in place to ensure competition advantages of the companies who implement them. Companies like Canon do not want to see a rival bring a camera in the market using their own format and/or proprietary software to process their raw images. And this is the main obstacle for open sourcing their formats.

Nevertheless, what the end user may or may not do with the raw files generated with a legally obtained camera, is not Canon's concern and it can't be, because any files generated with a camera are always the property of the camera's end user and not the manufacturer's.

I am not sure to what Dan referred to when said "Patents don't work that way!"
but legal precedent has shown that software patents work this way: if I loose money from you using my patent, then I sue you  and lawyers come into play, if I don't loose money or even benefit from your work, then I don't care, nevertheless I keep my patents in place.
So the question about foreseeable legal troubles boils down to: was exiv2 or darktable or anybody else faced legal trouble for reading Canon's CR2 raw format? (yes, there patents for that too).

I am not a lawyer but following @johnny-bit and @clanmills initiative I discussed this issue with one. The above comments can be considered a short summary of this discussion.

**Comment by clanmills** [comment]
The exiv2 project has no money at all.  None.  Not a penny.  So why would Canon sue somebody with no money?  And if the project has no money, Canon would find it difficult to explain in court why reading metadata (written to a public standard) caused them to loose money.  As I have pointed out, we could find the metadata by hunting for the markers.

After I started to work on v0.27.3, my wife said "Am I becoming an Exiv2 widow again?  You're in your office 12 hours every day.  When is this going to stop?".  So when Dan and Jens started being so negative, I thought:  "I don't need to spend 4 weeks developing and testing ISOBMFF.".

I'm not going to do anything at the moment.  Exiv2 v0.27.3 RC2 was published as scheduled last Sunday on 2020-05-31.  I hope there will be no code changes and it will be released as 0.27.3 on 2020-06-30.

If no-one else volunteers, I'll reopen #1066 for the next release (0.27.4 or 0.28).  However, I hope somebody will undertake this project.  I want to pursue other hobbies such as playing the Euphonium and Piano, running, the garden and spending time with my wife.

![MusicRoom](https://user-images.githubusercontent.com/529982/83955753-6e51ee00-a84e-11ea-9a16-47fef8fec5df.jpg)

**Comment by pr0m1th3as** [comment]
Apparently they wouldn't, but I guess we will have to wait and see what's next. I only wish I had the necessary skills to get involved in a more productive manner other than expressing my support.

On the bright side, your music room makes a clear statement about pursuing other hobbies. Although euphonium, running and gardening are not my things, the piano is a serious temptation :+1: and reasonably deserves for some daily attention, not to mention the wife of course.

**Comment by clanmills** [comment]
There's another reason that I forgot to mention.  I'd like to solve the "Lens Recognition" problem.  I'm going to make a proposal to the community at LGM in Rennes in May 2021.  https://github.com/Exiv2/exiv2/issues/1202#issuecomment-626142229

In order to work on M2Lscript (pronounced **MillsScript**), I have to escape from Exiv2 to give me the time.  I hope the community will find somebody to take on the maintenance of Exiv2.  I intend to run a work-shop at LGM based on my book.  I'm doing all I can to ensure the long-term success of Exiv2.

**Comment by D4N** [comment]
Robin Mills <notifications@github.com> writes:

> The exiv2 project has no money at all.  None.  Not a penny.  So why
> would Canon sue somebody with no money?

It's not about us, it is about the consumers of the exiv2
library. Because they might not have the luxury of a lawyer being able
to tell them that continuing to use libexiv2 is ok. Or their
lawyer/legal advisor tells them that it is too risky.

Given that a big chunk of the Linux Desktop stack depends on exiv2 makes
me feel a little uneasy about supporting patent encumbered formats, as
this is nothing that can be solved via compile time switches.

**Comment by johnny-bit** [comment]
@D4N - There is literally nothing legally speaking in the way of loading any file on your disk into memory, reading bytes from it and interpreting those. be it heif, mp3, jpg2k or whatever. Also if data is available within file in know standard, there's nothing legally barring you from interpreting that data according to that standard (meaning exif, xmp etc...).

As I mentioned - you are stomping on legal minefield if you try to decrypt encrypted data. You are violating law if you're using unlicenced proprietary decoder to read encoded data. You are threading near some borders if you do cleanroom-analysis of data within file and write your own decoder.

I assume than none of the above matches description of what's needed to read exif data from ISOBMFF container.

Legal-expertise wise: I asked my legal advisor for legal advice and it's cost. I'll setup a meeting with him and ask for cost of legal expertise. I can cover at least part of the cost associated. I also would be needing description of way of ow exiv2 reads files, does it use any proprietary software or any "problematic" parts I've mentioned above.

**Comment by pr0m1th3as** [comment]
@D4N Nikon's NEF is a proprietary format, Canon's CRW & CR2 are proprietary formats too, so could you clarify what's the difference with Canon's CR3? They all have patents filed to maintain advantage over their competitors. Even DNG has a patent license called "Digital Negative (DNG) Specification Patent License" although they 've stated about 5 years later (in 2009) that "there were no known intellectual property encumbrances or license requirements for DNG". (sourced from wikipedia and Adobe's website)

AFAIK if one runs into trouble for reading proprietary raw formats, then this trouble should be here onto all of us very long ago. If I remember correctly dcraw started reading such files by reverse engineering them almost two decades ago. To my knowledge, there has been no issue with any opensource software reading proprietary raw files all these years.

On the other hand, if your concern arise from the latest developments with EU legislation and the rumors/concerns that could potentially harm other open source projects like VLC for reading other media formats (audio or video), this is an entirely different story. Regarding VLC and the like, there are legal grounds because those files are (potentially) not legally acquired or/and they still have certain restrictions such as CDs, tapes, vinyl etc. for example you need extra permission to publicly play a music CD even if you bought lawfully. And there is a lot concern about online piracy and all that. Most importantly, when you buy some music on a CD and mp3 file, you own the medium not the intellectual content it is saved in there.

But this is never the case with our camera's raw files, any content saved in a raw format with a camera I take a picture with belongs to me including any copyrights I am willing to introduce to my photos. This the main reason why companies like Canon, Nikon etc never bothered coming after you or every other person like Robin who have devoted significant time into making these formats accessible to the open source community. The other reason is explained in my previous post (nothing to gain). They still charge (possibly) Adobe for some readily available API so they can include CR£ support in their latest LR version, which they charge end users for a lot of money, and they do this because they can (Adobe don't mind some premium for faster support for their software, they take for free from the opensource community a lot more, which is under LR's hood anyway). But you can add this to the reasons why Canon and the like are quite hesitant to provide a free API for reading their raw files.

**Comment by 1div0** [comment]
* please read this https://www.linkedin.com/pulse/future-without-mpeg-leonardo-chiariglione/

**Comment by pr0m1th3as** [comment]
>     * please read this https://www.linkedin.com/pulse/future-without-mpeg-leonardo-chiariglione/

I am not sure how this directly relates to the matter of this thread. Quite interesting reading though.

As for the purposes of reverse engineering and reading a proprietary format for interoperability purposes, the following US Law makes it pretty clear IMO and also explains why StarOffice, OpenOffice, & LibreOffice had no legal troubles reading MS Word formats before Microsoft's Open Specification Promise over a decade ago. Not to mention about dcraw, exiftools, exiv2, darktable, etc., which are our concern in this thread.

Sec. 103(f) of the DMCA (17 U.S.C. § 1201 (f)) says that a person who is in legal possession of a program, is permitted to reverse-engineer and circumvent its protection if this is necessary in order to achieve "interoperability" — a term broadly covering other devices and programs being able to interact with it, make use of it, and to use and transfer data to and from it, in useful ways. A limited exemption exists that allows the knowledge thus gained to be shared and used for interoperability purposes.
source: https://www.law.cornell.edu/uscode/text/17/1201

**Comment by clanmills** [comment]
Topic: Exiv2 and ISOBMFF Support
Time: Jun 14, 2020 13:00 London

Join Zoom Meeting
https://us02web.zoom.us/j/82136730279?pwd=M3hCbll4cWN3ellJd2pCZkxjVEx3Zz09

Meeting ID: 821 3673 0279
Password: 1fDNUV

**Comment by boardhead** [comment]
> I want to pursue other hobbies such as playing the Euphonium and Piano, running, the garden and spending time with my wife.

Hey.  Euphonium was my instrument too, but I haven't played since high school. :)

**Comment by clanmills** [comment]
I played Trumpet in the school orchestra and haven't played it since.  I decided to move down an octave and delighted by the experience.  I got the instrument in September 2018 and expect to pass Grade-6 in November.  I attend a "virtual" trumpet school in Australia.  The teacher gives us a video/lesson every day on Facebook for $30/month! (20 in the class).  I have a brass tutor who comes to my house every other Wednesday.  Since lockdown, we meet in a virtual class room.

I've played piano all my life.  We purchased a digital piano last year.  It's amazing.  It's beautiful to play.  It'll accompany me when I can get a midi file.  I purchased ScanScore which does scan (or PDF) to music.  So, if I can't find a midi file, ScanScore can "read the dots.".

Have a look at my book.  It's coming on well:  https://clanmills.com/exiv2/book/

In particular, please read this:  https://clanmills.com/exiv2/book/#future

**Comment by johnny-bit** [comment]
AVIF is an open standard, CR3 is/should be freely readable. JPEG2000 is already in the lib. only issue: HEIF. Still, reading bytes from file and interpreting known values doesn't touch file patents that I know of.

Using introspection and dynamic linking would've in edge cases circumvent any problems related to using totally free lib in patent minefield. Same thing would be if it were filetype libraries connection to exiv2 not other way around and announcing things they can provide. Then it'd be simple wine-brick story, only in "don't install libPatentedFileFormat along with exiv2 or you might've been able to read exif, xmp and ICMP data from PatentedFileFormat".

**Comment by clanmills** [comment]
@johnny-bit You're right. If we use a library which has code that infringes the patent, we could be in trouble, even if we don't use the part of the library that infringes the patent.

I'm also mistaken about having no libheif code.  We have to know the interface (a `.i` file in COM) to use introspection.  So we would have libheif code in Exiv2.  However, it may be possible to only include that part of the interface that we really need.  That may be possible (or impossible).

Incidentally, @cgilles submitted a PR today to link libheif #1233.  It only involves changing about 25 existing files and adding 53 new files.  The patch has 130,000 lines and cannot be applied to 'master'.  The PR is only as simple as this because it is readonly, and doesn't include any test code at all.  Clearly it's ready to ship (not).

We could return to my idea to write as little code as it takes to read the file and locate the 4 blocks of interest.  I was originally looking at Dave Coffman's parse.c which reads all manner of files in 1000 lines of C.

**Comment by boardhead** [comment]
> Have a look at my book. It's coming on well: https://clanmills.com/exiv2/book/

Holy cow!  This book looks like it will be amazingly comprehensive.  Great work so far!!

**Comment by cgilles** [comment]
I don't know with github PR make a garbage.

this is a "git diff HEAD > heif.patch.txt". 22 Kb file with limited in-deep code change. It very simple in fact...

[heif.patch.txt](https://github.com/Exiv2/exiv2/files/4754810/heif.patch.txt)

**Comment by cgilles** [comment]
Patch V2. I forget a block change in a CMakeLists.txt script:

[heifv2.patch.txt](https://github.com/Exiv2/exiv2/files/4754859/heifv2.patch.txt)

For an HEIF  unit test, it's very simple. Take an HEIF image taken with an iphone for ex, which contain metadata an try to use exiv2 CLI tool to display Exif|Iptc|Xmp.

libheif parse file to extract metadata chunk and pass byte-array as well to exiv2 core. No more. The rest is standard Exiv2 code unchanged...

This is a sample HEIF file with metadata:


[pict0027.zip](https://github.com/Exiv2/exiv2/files/4754864/pict0027.zip)

**Comment by paperdigits** [comment]
@D4N and @phako : 

Can you please outline the steps necessary, in your opinions, to get support merged into exiv2? 

You say you're worried about the consumers of the exiv2 library, and the linux desktop at large, but *we* are the consumers of the exiv2 library-- darktable, digiKam, elementary photos, GNOME Photos, etc etc.

**Comment by phako** [comment]
> You say you're worried about the consumers of the exiv2 library, and the linux desktop at large, but _we_ are the consumers of the exiv2 library-- darktable, digiKam, elementary photos, GNOME Photos, etc etc.

All I said was give people (i.e. individuals and distributions) who care about such things as non-free and potentially patent-proned formats an opt out. Nothing more. Because doing that later is awful and much easier if you design the whole feature with it from the start.

**Comment by paperdigits** [comment]
@phako thanks!

We'll wait on the reply of @D4N

**Comment by debarshiray** [comment]
As the upstream author of GNOME Photos, a GNOME contributor in general, and a Fedora contributor, I totally agree with @phako here.

One approach that's used by some frameworks (eg., GStreamer) is to use loadable (via `dlopen`) shared objects (or plugins) for codec support. That way, distributors (like Fedora) that cannot ship patent encumbered code can remove those plugins, but third-party repositories can provide the patent encumbered codecs for users who are able to use them.

**Comment by clanmills** [comment]
Thanks @debarshiray.  You have sunk this project beyond hope.  All the work that I have invested into ISOBMFF support for Exiv2 has been wasted.  And this has been achieved:

1) Without anybody explaining which patents might be infringed.
2) Without any complaint from any patent holder.
3) Even although other open-source projects have decoded ISOBMFF files without legal consequences.
4) Without including code in Exiv2 from any other project.
5) Despite the fact that we only need to modify jp2image.cpp which has been in Exiv2 for about 10 years.

There's one silver lining.  I had agreed to get a legal opinion before publishing any code.  I'm glad I didn't use my legal resources as it's clear that no solution to supporting ISOBMFF in Exiv2 will be accepted by some members of the community.

**Comment by clanmills** [comment]
I'm closing this issue as there is now no possibility of adding support for ISOBMFF Files.

**Comment by phako** [comment]
Sorry I do not understand what made you get this understanding from @debarshiray 's comment

**Comment by Filou83** [comment]
> 
> 
> I'm closing this issue as there is now no possibility of adding support for ISOBMFF Files.

Could you please re-open the issue? We need support for HEIF files in Exiv2! Not supporting it is a Disservice for the open source community by blocking modern image formats, because the usage of these formats is increasing more and more.

**Comment by debarshiray** [comment]
I apologize if it came across as too grim. Software patents are considered evil for a reason, after all -- they do create a lot of needless pain for everybody involved.

Usually, a project without any money doesn't need to worry because since they have no money there's no incentive to sue them. However, a project might become [a victim to it's own success](https://www.gnome.org/news/2019/10/gnome-files-defense-against-patent-troll/) in raising funds, and attract legal action.

Then there's the issue of jurisdiction. If you are in a part of the world that doesn't recognize software patents, then you can get away with ignoring them, else you cannot.

One big problem with patents is that it's not just the Exiv2 project that can be sued, even downstream distributors become liable. So, depending on which part of the world you are, and what your finances look like, your level of paranoia can vary. eg., Fedora is very wary of patents because it's sponsored by an entity (ie., Red Hat) that's registered in the United States of America, whereas others might or [might not be](https://rpmfusion.org/).

It's also worth noting that it's not advisable to discuss specific patents in public because of the likelihood of [treble damages](https://en.wikipedia.org/wiki/Treble_damages).

However, it is very widely known in free software circles that file formats developed by the [Moving Picture Experts Group](https://en.wikipedia.org/wiki/Moving_Picture_Experts_Group) tend to be heavily encumbered by patents. eg., H.264, H.265 or HEVC, AAC, etc.. If you are looking to get in touch with people with years of experience dealing with multimedia formats in the trenches, then I'd suggest getting in touch with the GStreamer maintainers.

Having said that, patents do expire. And by carefully navigating the patent landscape, with proper legal advice, one could over time start shipping code that they previously couldn't. eg., see how Fedora started enabling [MP3](https://fedoramagazine.org/full-mp3-support-coming-soon-to-fedora/), [AC3 or A52](https://blogs.gnome.org/uraeus/2017/03/22/another-media-codec-on-the-way/) and [a sub-set of AAC](https://blogs.gnome.org/uraeus/2017/10/12/aac-support-will-be-available-in-fedora-workstation-27/).

Given all this, a plugin framework for file format support is a practical solution that some projects employ. If some distributor is unable to ship some code, then leave those plugins out, but others can add them back later, if they want to. The upstream project also gets the opportunity to clearly spell out which parts of the code might be encumbered by patents.

Of course, plugins aren't mandatory. If the code is baked into a single library, then distributors can always patch out the parts that they don't want to carry. However, this is less elegant because it's not easy to restore the removed code, and the downstream patch might introduce bugs or remove too much or too little.

**Comment by clanmills** [comment]
The issue of providing ISOBMFF support in Exiv2 cannot be solved.  Every solution proposed has been rejected.  We have discussed _and decided against_ the plugin mechanism.  https://github.com/Exiv2/exiv2/issues/1229#issuecomment-641509835

Equally to the point.  Image Patents concern encoders and decoders.   My ISOBMFF reader (in my book) requires no encoder or decoder.

I have given up.  Defeated by the community.

**Comment by debarshiray** [comment]
> The issue of providing ISOBMFF support in Exiv2 cannot be solved.
> Every solution proposed has been rejected. We have discussed and
> decided against the plugin mechanism. #1229 (comment)

I couldn't figure out how that comment is related to the issue of plugins, but I am sure that you have good reasons for arriving at that decision.

> Equally to the point. Image Patents concern encoders and decoders.
> My ISOBMFF reader (in my book) requires no encoder or decoder.
>
> I have given up. Defeated by the community.

You are not going to able to solve complicated legal issues in the public via an issue tracker. The peanut gallery won't be able to add anything useful to the discussion.

I, personally, am not an expert in multimedia formats to be able to decide in an instant whether the code in question may or may not encroach on any patents. However, both @phako and I have been in the trenches long enough to know that anything coming out of MPEG can be suspicious and should be treated carefully.

Therefore, instead of arguing this in public, I'd suggest talking to people with decades of experience on this topic - both technical and philosophical. My recommendation would be to start with the GStreamer maintainers. If you want, I can help with the introductions.

**Comment by clanmills** [comment]
I have given up.  Further discussion is pointless.

**Comment by debarshiray** [comment]
> I had agreed to get a legal opinion before publishing any code. I'm
> glad I didn't use my legal resources as it's clear that no solution to
> supporting ISOBMFF in Exiv2 will be accepted by some members
> of the community.

The free software community does have well-respected and qualified lawyers who volunteer their expertise pro bono for such matters.

But who knows, if you spoke to the right people, maybe it wouldn't go that far. Maybe your opinion that your code doesn't touch any patents would be validated and everything would be good.

**Comment by clanmills** [comment]
I believe it is legal.  My original plan was to get the code ready, seek legal advice, then publish the code in Exiv2 v0.27.3 RC2.  When I proposed that course of action, I was told "you don't know how patents work." - which is of course why I proposed to get legal advice.

If I had proceeded with my plan, Exiv2 v0.27.3 would probably now have ISOBMFF support.  Instead, Exiv2 v0.27.3 has shipped without ISOBMFF support and I've had to deal with hundreds of emails about this subject.

I need a vacation.  I don't know if I even have the energy to finish my book.

**Comment by clanmills** [comment]
Thank you to @debarshiray for offering to provide an introduction to legal resources.  I appreciate your offer to help.

And thank you to everybody who has contributed to this discussion.  From the outset of #1066, we all recognised the possibility of legal issues.  https://github.com/Exiv2/exiv2/issues/1066#issuecomment-605454862

Since Exiv2 v0.27.3 RC2 was released on 2021-05-31 without ISOBMFF support, I've been deluged with emails asking why this has been blocked on legal grounds.  I've allowed this conversation to consume too much of my time.

That's how it is.

I'm going to get the book finished.  If somebody would like to undertake the work to get a legal opinion about this, the road may be open to adding the code from my book to Exiv2 in 2021.   However, I'm not offering to work with the legal people, nor to port the code from my book to Exiv2.

I want to complete the book in 2020 and retire before my 70th birthday on 2021-01-18.

I'm not going close the thread because the subject is not resolved.  The valuable input that everybody has made on this thread is a good starting point for a future maintainer to consider how to satisfy the community's strong desire to have ISOBMFF support in Exiv2.

**Comment by novomesk** [comment]
In my programs I use following workarounds:

When reading metadata I let the libavif or libheif to parse the AVIF or HEIF file, then I has the buffer with metadata to Exiv2.

Before saving AVIF or HEIF files, I prepare metadata with Exiv2's help. The I use `Exiv2::ExifParser::encode` and pass returned data from Blob to libavif or libheif to save the file.

Updating metadata in existing AVIF or HEIF files is more challenging to implement for me but so far I can live without it.

**Comment by dhoulder** [comment]
If no other legal advice is forthcoming, the Software Freedom Law Center may be able to help. See
https://www.softwarefreedom.org/about/contact/
I'd be happy to contact them and outline the situation.

**Comment by paperdigits** [comment]
Thank you for the offer.

We're really just waiting to hear the opinion of @D4N.

**Comment by clanmills** [comment]
Thank you, David.  I would very happy if you could discuss this with the Software Freedom Law Center.

Let me outline the position as I see it:

**What do we want to do?**

There have been numerous requests to the Exiv2 project to read/write/modify ISOBMFF files.  This include Apple's HEIF, Canon's CR3 and the AVIF format. #1066.  ISOBMFF format is defined in the following standard.  https://www.iso.org/standard/78321.html

There are 4 types of metadata of interest to exiv2 and they are Exif, IPTC, ICC, XMP.  These data blocks are embedded in the ISOBMFF as sub boxes of the "meta" box.  They are unencrypted and stored as specified in the ISOBMFF Standard.

Exiv2 has provided read/write support for JPEG 2000 files for about 10 years.  My book **Image Metadata _and_ Exiv2 Architecture** documents how that can be extended to achieve our goal:  http://clanmills.com/exiv2/book/#ISOBMFF

**Why do we believe this is legal?**

1. Other open-source projects read ISOBMFF files without legal consequences.
2. No code from other projects has been included in Exiv2.
3. The implementation is achieved by modifying jp2image.cpp which has been in Exiv2 for about 10 years.
4. No complaint has been received from any patent holder.
5. The ISOBMFF format is based on MP4 and is now more that 25 years old.
6. Image format patents deal with image encryption.  The metadata is not encrypted.
7. Exiv2 has no code to read or modify the image.

**Why do we believe this could be illegal?**

The case for this being illegal is historical.  In the 1990s there were patent wars concerning several graphic image formats.

The ISOBMFF specification states:  _Attention is drawn to the possibility that some of the elements of this document may be the subject of patent rights. ISO and IEC shall not be held responsible for identifying any or all such patent rights._

@debarshiray has expressed the concern _I have been in the trenches long enough to know that anything coming out of MPEG can be suspicious and should be treated carefully._.  https://github.com/Exiv2/exiv2/issues/1229#issuecomment-691213222

In response to my assertion that it is not illegal to read an ISO Standard file, @phako has stated _I think ISO/IEC 13818 (MPEG-2) and ISO/IEC 14496-10 (H.264/MPEG4-10) tell a different story._  https://github.com/Exiv2/exiv2/issues/1066#issuecomment-609879359

@D4N has stated _It is about the consumers of the exiv2 library. Because they might not have the luxury of a lawyer being able to tell them that continuing to use libexiv2 is ok. Or their lawyer/legal advisor tells them that it is too risky.  Given that a big chunk of the Linux Desktop stack depends on exiv2 makes me feel a little uneasy about supporting patent encumbered formats, as this is nothing that can be solved via compile time switches._  https://github.com/Exiv2/exiv2/issues/1066#issuecomment-589929474

**Alternative solutions**

Several PRs have been provided by users to provide this support in Exiv2.  None have been merged as they are incomplete and rely on third party libraries.

There has also been discussion about using plugins and introspection to access external code which will read the file.

None of those solutions address the legal situation as they are different ways of "breaking into the file".  If accessing the metadata in an ISOBMFF files is a patent infringement, it makes no difference how access is achieved.

**What's the desired output?**

I hope this can be resolved to a general statement such as "It is legal to read unencrypted metadata in an ISOBMFF file.".  Such a resolution would not be tied to the implementation in my book and the project could use any implementation to provide Exiv2 users with support for ISOBMFF.  Additionally, such a resolution would absolve all existing and future metadata processing code from further discussion of this issue.

**Comment by dhoulder** [comment]
Thanks Robin. That's an excellent summary. I've just sent an email to help@softwarefreedom.org
Text attached.
[SFLC-request.txt](https://github.com/Exiv2/exiv2/files/5347647/SFLC-request.txt)

**Comment by dhoulder** [comment]
No news from the Software Freedom Law Center yet. A few hours after sending the initial request I received a confirmation email from them which read, in part..
> we need to make sure that you understand that your email to us does not create an attorney-client relationship with us and any information you send us will not be considered confidential or privileged. If you understand that, just reply to this message by keeping the text of this paragraph and adding "Understood" and we will respond to your email shortly.

I replied as directed on 9 October, and on 16 October I sent a followup message to check that they had received my reply. No response yet.

**Comment by clanmills** [comment]
@dhoulder Thanks, David. It's disappointing that SFLC have given not yet provided any meaningful feedback.

@debarshiray suggested https://github.com/Exiv2/exiv2/issues/1229#issuecomment-691216937 and 

> My recommendation would be to start with the GStreamer maintainers. If you want, I can help with the introductions.

Perhaps we have to approach a different group for legal support.

**Comment by paperdigits** [comment]
Hopefully @D4N can provide an opinion soon so we can get moving.

**Comment by dhoulder** [comment]
@debarshiray I'd be happy to start talking to the GStreamer maintainers

**Comment by dhoulder** [comment]
I just received a response from the SFLC helpline:

> Our apology for the inordinate delay. We are pressed for resources these days to address all matters. We would be delighted to offer assistance. Please let us  know  if you have time for a  phone call this week. Please send us time slots  for 30 min on ET hours that may work for you and we will schedule a  consultation.

It might be good to get one or two others in on the call if that's possible. I'm 16 hours ahead of ET, but I can probably do 08:00 ET (13:00 UTC) or 17:00 ET (22:00 UTC). How about Thursday 5 November? I can do other days if that's inconvenient.

**Comment by clanmills** [comment]
4pmEST Thursday 2020-11-04 = 21:00GMT = 08:00AEDT Friday 2020-11-05? (+/-1 hour or 30 minutes).

Good News, David.  I'm happy to join you.  I hope this doesn't have to be a "phone" call and we set this up on Zoom.  Can you ask permission to record to record the call.

**Comment by dhoulder** [comment]
> 4pmEST Thursday 2020-11-04 = 21:00GMT = 08:00AEDT Friday 2020-11-05? (+/-1 hour or 30 minutes).

Sounds good.  I'll organise that.

Yes, Zoom would be preferable. I'll ask about that and about recording.

**Comment by tp-m** [comment]
Hi,

GStreamer (multimedia framework) developer here.

I'm not a laywer, nor do I play one on TV, so I don't really have any special
insights or knowledge. There are a lot of different views on this topic in
terms of what can be patented (and where), what should be allowed to be
patented or not, whether any of it matters for FLOSS projects, etc. Like 
everyone else, I have opinions on these subjects, but they really don't matter
and even if you knew what they were you should probably pay no attention to
them.

It's good to be aware of and pay attention to these kind of issues, especially
if you are a widely used library such as exiv2 that wants to make sure it
doesn't create issues for its users and distributors, so it's good to have
this discussion.

However, the truth of the matter is that you will be hard-pressed to find
someone who will provide you with definite legal advice or legal certainty
that this is definitely 100% okay, and not just wherever you're based but
also in all other markets or jurisdictions on this planet.

This means that as a project/community you will probably not get around to
applying some common sense (for lack of a better word) if you want to move
forward.

I mean, we're talking about metadata extraction of a really simplistic
container format here, that's not much different than EXIF or XML really.

It's usually image/video decoding that's more problematic (esp. H.264/AVC and
H.265/HEVC video), not so much the containers.

If you want to make sure there's unlikely to be any problems for third parties
with whatever you're planning on doing, I think the best thing you can do is
to look at what companies/projects who take this kind of thing seriously have
decided. The gold standard for this in the FLOSS world is RedHat, so if
something ships out of the box on Fedora/RHEL you can be fairly sure it's
fine and is unlikely to cause problems for anyone.

And if someone did try to cause problems that might be a prime case for the
GNOME patent troll defense fund ;)

Anyway, what does that mean for the issue at hand:

1. Would be a problem for users/distributors if you included code in exiv2 that
   parses an isobmff file structure to extract and/or write HEIF/AVIF/CR3
   metadata and/or compressed image data (without decoding/encoding the image)?
   The answer to this question is probably a resounding "no, don't worry about it".
   Not because I say so, but because it's clearly not regarded as problematic by
   distributors who take such issues seriously such as e.g. RedHat: GStreamer has
   been shipping a fully-featured isobmff/mp4/quicktime demuxer and muxer in its
   gst-plugins-good plugin set since 2011. Fedora ships that out of the box, as
   well as libmp4v2 ("Library for working with files using the mp4 container
   format") and libavif and probably lots of other software that does exactly
   that. It is usually the decoding or encoding of the actual image/video data
   that is considered problematic (depending on the format, of course).

2. Would it be a problem for distributors if you relied on third party libraries
   such as libheif or libavif to extract/write such metadata? The answer to
   that is probably "it depends". libavif shouldn't be a problem, since as
   mentioned above that is already widely shipped. libheif on the other hand
   would be a problem for many distributors, because it depends on image
   decoding/encoding libraries such as libde265 and libx265 which are generally
   regarded as problematic from a patent point of view. Now that doesn't necessarily mean
   that you can't make exiv2 depend on libheif for this functionality, but if
   you do so you should really make it an optional dependency (i.e. provide
   a way for exiv2 to not use this library and not provide this functionality
   then). From a distro/user point of view that'd probably work technically,
   but it would be suboptimal because then some functionality can't be provided
   to all users. As such, including your own isobmff parser in exiv2 or relying
   on a library such as libmp4v2 that is widely shipped and does not rely on
   problematic image decoder/encoder libraries would probably provide a better
   user experience overall, because everyone gets full functionality.

In response to the specific concerns mentioned above in https://github.com/Exiv2/exiv2/issues/1229#issuecomment-705350266:

> @debarshiray has expressed the concern: _I have been in the trenches long enough to know that anything coming out of MPEG can be suspicious and should be treated carefully._. [#1229 (comment)](https://github.com/Exiv2/exiv2/issues/1229#issuecomment-691213222)

It's good to raise awareness that this could be a problem, but I don't think there is any concrete evidence here or anywhere that anyone regards this particular issue (isobmff container parsing to extract metadata) as particularly problematic.

> In response to my assertion that it is not illegal to read an ISO Standard file, @phako said: _I think ISO/IEC 13818 (MPEG-2) and ISO/IEC 14496-10 (H.264/MPEG4-10) tell a different story._ [#1066 (comment)](https://github.com/Exiv2/exiv2/issues/1066#issuecomment-609879359)

Not wanting to put words in @phako's mouth, but I think that just means that the fact that something is an ISO standard doesn't mean anything one way or another, i.e. ISO 14496-10 H264/AVC decoding might be problematic even though it's an ISO standard, but that doesn't mean in turn that anything that's an ISO standard is problematic, just that one has to look at each case individually.

> @D4N has stated: _It is about the consumers of the exiv2 library. Because they might not have the luxury of a lawyer being able to tell them that continuing to use libexiv2 is ok. Or their lawyer/legal advisor tells them that it is too risky. Given that a big chunk of the Linux Desktop stack depends on exiv2 makes me feel a little uneasy about supporting patent encumbered formats, as this is nothing that can be solved via compile time switches._ [#1066 (comment)](https://github.com/Exiv2/exiv2/issues/1066#issuecomment-589929474)

As I said above, if RedHat (and Ubuntu and Suse etc.) have decided that it's perfectly fine to ship other code that does this in the base distro, that probably means that it's perfectly fine for all other consumers of the exiv2 library as well.

**Comment by clanmills** [comment]
Thanks for this input.  @dhoulder and I have an appointment to meet with somebody from SFLC shortly.  The meeting was requested for 2021-11-05 21:00 GMT.  As the meeting hasn't been confirmed, we're expecting a delay. 

I expect the result of the discussion with SFLC will be vague.  Although I've defined my desired outcome, it's more likely to be something vague such as "it's probably OK, but ..... ".

I'll update the thread when there is something to report.

**Comment by clanmills** [comment]
The meeting with SFLC will take place on Zoom at 21:00GMT 2020-11-10.  Currently, @dhoulder and I will attend and two people from SFLC in New York City.  If anybody wishes to attend please let me know.  I will report here after the meeting.

**Comment by clanmills** [comment]
Today, in the course of writing about XMPsdk my book, I was digging in the Adobe XMPsdk which Exiv2 has been shipping for about 15 years.  

In the latest XMPsdk (available on GitHub) DumpISOBoxes() in samples/course/common/DumpFile.cpp has the following:

```cpp
else if (isoPath == "meta/") // for HEIF
{
	digestISOFullBoxExtension(file, isoPath, remainingSize, version, flags);
	DumpISOBoxes(file, remainingSize, isoPath);
	break;
}
```

Adobe's code is BSD licensed.   

I haven't studied this code, however it appears similar to my code in tvisitor.cpp in my book.

As a retired Adobe Engineer, I am 100% certain that this code has been audited and approved by Adobe Legal.  In addition to the DigiDNA open source in ISOBMFF Explorer, we now have evidence that Adobe also ship open source that reads the ISOBMFF file.

**Comment by KristijanZic** [comment]
How did the meeting with SFLC go?

**Comment by clanmills** [comment]
It was very constructive.  I was unclear about the confidentiality of our discussion.  So, I've drafted a statement which is being discussed in private by email.  I hope an update will be posted in the next couple of days.

**Comment by KristijanZic** [comment]
Alright, thanks! I really hope the ISOBMFF support is now being green-lighted.

**Comment by dhoulder** [comment]
I concur with Robin — it was a very useful meeting. I'll post an update here in the next 2 or 3 days.

**Comment by dhoulder** [comment]
Robin and I have discussed what productive steps might be taken next, and think that in order to protect Exiv2 and its users, we should consider joining the Open Invention Network. See https://openinventionnetwork.com/

OIN is a consortium of major technology players including Redhat, Google, and IBM,  with over 3300 members all up. You can read about the benefits of joining OIN  at https://openinventionnetwork.com/about-us/member-benefits/
 
They have a license (https://openinventionnetwork.com/joining-oin/join-now/license-agreement/ ) which Exiv2 could sign which would commit us to:

-  Forfeit our right to sue other members for patent infringement.
-  Donate our patents for unconditional use by other members.

There's no fee for joing OIN. There are potential benefits in the event of any legal issue arising with Exiv2, possibly even with existing code, mostly due to the membership of some very large industry leaders.

Furthermore, they have a concept of a Linux System Definition, which includes core Linux and adjacent open source technologies. See https://openinventionnetwork.com/linux-system/

Note that currently included in the Linux System Definition are imaging-related packages such as gimp, gd, ImageMagick, and digikam, so I expect Exiv2 would be a prime candidate.

**Comment by clanmills** [comment]
Thank you @dhoulder for working with me on this issue.

On 2021-01-18 (9 weeks on Monday), I will retire.  My book will be complete.  I hope somebody will step forward to maintain Exiv2 and I assure them of my support and cooperation.  I hope my successor with get Exiv2 into OIN and the Linux System Definition.

Concerning ISOBMFF, here's my summary of the situation.

1.  The risk of patent infringement in ISOBMFF is the same as the risk in the current exiv2 library.
2.  In the investigation of this issue, no strong argument has surfaced to stop this development.
3.  Exiv2 users (both developers and end users) have strongly expressed support for this feature.  If we don’t provide this feature, somebody will fork the code.  As the community struggles to maintain the two branches of Exiv2, a second fork will cause confusion and duplication.

If somebody offers a PR, I will accept it provided:

1.  Nobody provides a strong reason to say STOP.
2. The PR builds on all supported platforms and passes the existing test suite.
3. The PR extends the test suite to test the newly supported formats.
4. The PR has a build switch to enable/disable the feature.
5. The PR does not require Exiv2 to call a library.
6. The PR Includes appropriate documentation updates.
7. The author of the PR agrees to support users and fix issues in the code for 2 years on both the 0.27-maintenance and master branches.

I've personally found working on Exiv2 to be interesting and mostly enjoyable.  It has been a surprise that more contributors haven't joined in the fun. On page 2 of the book, I acknowledge: _Exiv2 contributors (in alphabetical order): Abhinav, Alan, Andreas (both of them), Arnold, Ben, Gilles, Kevin, Leo, Leonardo, Mahesh, Micha&lstrok;, Mikayel, Milo&scaron;, Nehal, Neils, Phil, Rosen, Sridhar, Thomas, Tuan .... and others who have contributed to Exiv2._

I hope the C19 situation fades in spring and LGM goes ahead in Rennes in 2021.  I look forward to seeing you there.

**Comment by 1div0** [comment]
Thanks, Robin!

So I will reload and restart my efforts to get HEIF support into the Exiv2 and then darktable with full metadata I/O.

Cheers!

Happy WE.

**Comment by johnny-bit** [comment]
@1div0 - that's good to hear. I myself wanted to start working on pure C++ implementation making sure it doesn't need/call external libraries as per #5 of Robin's requirements for PR :)

**Comment by clanmills** [comment]
The tvisitor.cpp code in my book (svn://dev.exiv2.org/svn/team/book) reads the metadata in JP2, AVIF, HEIF and CR3.  I have documented an explained in the book how it works.

The tvisitor.cpp code is less than 3000 lines and decodes Exif, IPTC, XMP and ICC metadata in every format supported by EXIV2, plus ISOBMFF and BigTiff.  It's a single file of code.  If you wish to decode PNG you need to link libz.  My aim is to demystify how metadata is encoded in all the common image formats.

**Comment by clanmills** [comment]
@1div0 Thanks for your email.  My reply bounced on your mail server.

I've fixed the "double free" in class CRW on Linux.  I've also fixed a Linux compiler warning in dmpf.cpp

I've look at your "NEOWISE" image and see:

```
548 rmills@rmillsmm-local:~/gnu/exiv2/team/book/build $ ./tvisitor -pR ~/Downloads/20200717_221452.avif 
STRUCTURE OF JP2 (avif) FILE (MM): /Users/rmills/Downloads/20200717_221452.avif
 address |   length |  box | uuid | data
       0 |       24 | ftyp |      | avif____avifmif1miafMA1B 97 118 105 102 0 0 0 0 97 118 105 102 109 105 102 49 109 105 97 102 77 65 49 66
      32 |     9336 | meta |      | _______(hdlr________pict____________liba 0 0 0 0 0 0 0 40 104 100 108 114 0 0 0 0 0 0 0 0 112 105 99 116 0 0 0 0 0 0 0 0 0 0 0 0 108 105 98 97
  STRUCTURE OF JP2 FILE (MM): /Users/rmills/Downloads/20200717_221452.avif:44->9332
         0 |       32 | hdlr |      | ________pict____________libavif_ 0 0 0 0 0 0 0 0 112 105 99 116 0 0 0 0 0 0 0 0 0 0 0 0 108 105 98 97 118 105 102 0
        40 |        6 | pitm |      | _____. 0 0 0 0 0 1
        54 |       36 | iloc |      | ____D__._.___.__$._.q:_.___._...__I. 0 0 0 0 68 0 0 2 0 1 0 0 0 1 0 0 36 168 0 15 113 58 0 2 0 0 0 1 0 15 149 226 0 0 73 234
        62 |       14 |  ext |    1 |   9384,1012026
        76 |       14 |  ext |    2 | 1021410, 18922
        98 |       57 | iinf |      | _____.___.infe.____.__av01Color____.infe 0 0 0 0 0 2 0 0 0 26 105 110 102 101 2 0 0 0 0 1 0 0 97 118 48 49 67 111 108 111 114 0 0 0 0 25 105 110 102 101
    STRUCTURE OF JP2 FILE (MM): /Users/rmills/Downloads/20200717_221452.avif:44->9332:112->51
           0 |       18 | infe |      | .____.__av01Color_ 2 0 0 0 0 1 0 0 97 118 48 49 67 111 108 111 114 0
          26 |       17 | infe |      | .____.__ExifExif_ 2 0 0 0 0 2 0 0 69 120 105 102 69 120 105 102 0
    END: /Users/rmills/Downloads/20200717_221452.avif:44->9332:112->51
       163 |       18 | iref |      | _______.cdsc_._._. 0 0 0 0 0 0 0 14 99 100 115 99 0 2 0 1 0 1
       189 |     9135 | iprp |      | __#.ipco___.ispe______..__..___.pixi____ 0 0 35 152 105 112 99 111 0 0 0 20 105 115 112 101 0 0 0 0 0 0 15 160 0 0 23 128 0 0 0 16 112 105 120 105 0 0 0 0
    STRUCTURE OF JP2 FILE (MM): /Users/rmills/Downloads/20200717_221452.avif:44->9332:197->9135
           0 |     9104 | ipco |      | ___.ispe______..__..___.pixi____....___. 0 0 0 20 105 115 112 101 0 0 0 0 0 0 15 160 0 0 23 128 0 0 0 16 112 105 120 105 0 0 0 0 3 8 8 8 0 0 0 12
      STRUCTURE OF JP2 FILE (MM): /Users/rmills/Downloads/20200717_221452.avif:44->9332:197->9135:8->9104
             0 |       12 | ispe |      | ______..__.. 0 0 0 0 0 0 15 160 0 0 23 128
            20 |        8 | pixi |      | ____.... 0 0 0 0 3 8 8 8
            36 |        4 | av1C |      | ..._ 129 31 12 0
            48 |     9048 | colr |      | prof__#Tlcms..__mntrRGB XYZ .._._._._,_; 112 114 111 102 0 0 35 84 108 99 109 115 2 16 0 0 109 110 116 114 82 71 66 32 88 89 90 32 7 228 0 7 0 19 0 9 0 44 0 59
      END: /Users/rmills/Downloads/20200717_221452.avif:44->9332:197->9135:8->9104
        9112 |       15 | ipma |      | _______._...... 0 0 0 0 0 0 0 1 0 1 4 1 2 131 4
    END: /Users/rmills/Downloads/20200717_221452.avif:44->9332:197->9135
  END: /Users/rmills/Downloads/20200717_221452.avif:44->9332
    9376 |  1030948 | mdat |      | ._....>~...@..d#..c...p..."."..=_'.....0 18 0 10 8 31 239 62 126 239 229 137 64 26 14 100 35 184 204 99 211 180 20 112 194 128 8 34 171 34 154 226 61 0 39 252 2 185 218 236 48
END: /Users/rmills/Downloads/20200717_221452.avif
549 rmills@rmillsmm-local:~/gnu/exiv2/team/book/build $ 
```

I don't know why you referred me to that image, however I'm sure it'll all make sense when we discuss it some more.  I'm looking forward to working with you - together we'll move some little mountains!

**Comment by alexvanderberkel** [comment]
hi everyone,

happy new year to all of you and I hope that all of you are safe and healthy. As the new Canon format CR3 is still an open issue for tools such as darktable, rawtherapee and many more, there was the idea to form a little task force from different projects in order to get this done. Within this the meeting we should discuss what are the blockers, e.g. legal questions / joining the open-invention network, what can be tackled by whom etc... 

I hope that the time slot on the weekend fits most people. 

_Topic:  Support for ISOBMFF Files (AVIF, HEIF, CR3)
Time: Jan 9, 2021 07:30 PM Amsterdam, Berlin, Rome, Stockholm, Vienna

Join Zoom Meeting
https://us05web.zoom.us/j/82168885673?pwd=K3pqRThERVNQK2JQR2ZweStwS2Nkdz09

Meeting ID: 821 6888 5673
Passcode: L1B3Hz_

Cheers,
Alex

**Comment by clanmills** [comment]
We have a horrible communications setup here.  There are different discussions about this taking place here, discuss.pixls.us, Wire, private email.  Can we discuss this here:

https://app.element.io/#/room/#exiv2-chat:matrix.org.  

@LeoHsiao1 Could you participate in this please.  I'm aiming to ship Exiv2 v0.27.4 with ISOBMFF support on 2021-04-30 and that will include your python test code (and the death of the bash scripts).

**Comment by LeoHsiao1** [comment]
@clanmills  
Thanks for your invitation, but I can't communicate orally in English.
Please arrange tasks for me through words. That's enough.

**Comment by clanmills** [comment]
Thanks, @LeoHsiao1

I think it's unlikely that you will be expected to do anything.  We will have to update the test suite of course, however that's unlikely to require your input.  I'll let you know if we need your help.

The good news here is that Exiv2 v0.27.4 is likely to ship at the end of April 2021 and will include your test code.  What happens to Exiv2 after v0.27.4 is not my concern and hope that the joint project with **darktable** will lead to a bright future for Exiv2.

**Comment by clanmills** [comment]
Agenda for the meeting:

[Exiv2-ISOBMFF-2021-01-09.pdf](https://github.com/Exiv2/exiv2/files/5790306/Exiv2-ISOBMFF-2021-01-09.pdf)

**Comment by clanmills** [comment]
Exiv2 0.27.4 Progress is being documented here:  https://github.com/Exiv2/exiv2/issues/1018#issuecomment-757979745

**Comment by clanmills** [comment]
Fixed.  #1475.  **Seven Engineers** have contributed to this.  @alexvanderberkel @LeoHsiao1 @hassec @clanmills @kmilos @piponazo @1div0.

And I must mention David Houlder for his help with the legal investigation.  And Gordon Laing for test images.  Pat and his folks at discuss.pixls.us who will set up a test image server.

By far the largest team of contributors in the history of the Exiv2 project.

**Comment by johnny-bit** [comment]
Thank you @clanmills and all contributors that made it possible!

**Comment by pr0m1th3as** [comment]
Big Thanks to @clanmills and everyone else involved in adding support for CR3 !!
